# DS-SF-36 | 03 | `pandas` | Assignment | Starter Code

## `pandas` and Bistro

You've just told one of your friend that you are taking a Data Science class.  (Yeah!)  Your friend is running a bistro, a small restaurant, serving moderately priced simple meals in a modest setting ([Wikipedia](https://en.wikipedia.org/wiki/Bistro)).  She collected over some period of time the following information of her patrons' visits.

| Variable's name | Its meaning |
|:---:|:---|
| `name` | Patron's first name |
| `gender` | Patron's gender |
| `is_smoker` | Whether the patron is smoking or not |
| `party` | Party's size |
| `check` | Check amount (\$) (after taxes but before tips) |
| `tip` | Tip (\$) that the patron added to the check |
| `day` | Week day of the visit |
| `time` | Rough time estimate of the visit |

In this assignment, we will be exploring this dataset using `pandas`.<sup>(*)</sup>

<sup>(*)</sup> this dataset was adapted from the `tips` dataset of the `seaborn` package (https://github.com/mwaskom/seaborn-data)

> ### Question 1.  Import `pandas` (as `pd`).

In [2]:
import os
# try and understand form Corey what does OS do?
import numpy as np
import pandas as pd

# TODO

pd.set_option('display.max_rows', 10)
pd.set_option('display.notebook_repr_html', True)
pd.set_option('display.max_columns', 10)

> ### Question 2.  Read the `dataset-03-bistro.csv` dataset.

In [6]:
# TODO
!pwd
#The present working directory is notebook whereas we have to navigate to datasets to be able to read the dataset
dframe = pd.read_csv(os.path.join('..','datasets','dataset-03-bistro.csv'))
#the dot dot sign here takes out of the directory to the previous directory so that you can get to the excel sheet from there
dframe

/Users/issingh/Documents/GitHub/DS-SF-36/classes/03/notebooks


day    time      name  gender  is_smoker  party  check   tip
0      Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1      Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66
2      Sunday  Dinner     Larry    Male      False      3  21.01  3.50
3      Sunday  Dinner    Joseph    Male      False      2  23.68  3.31
4      Sunday  Dinner    Janice  Female      False      4  24.59  3.61
..        ...     ...       ...     ...        ...    ...    ...   ...
239  Saturday  Dinner     Kevin    Male      False      3  29.03  5.92
240  Saturday  Dinner    Sandra  Female      False      2  27.18  2.00
241  Saturday  Dinner      Carl    Male      False      2  22.67  2.00
242  Saturday  Dinner       Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner    Brandi  Female      False      2  18.78  3.00

[244 rows x 8 columns]

In [10]:
type(dframe)
dframe.head
dframe.tail
dframe.columns


Index([u'day', u'time', u'name', u'gender', u'is_smoker', u'party', u'check',
       u'tip'],
      dtype='object')

> ### Question 3.  What is the class of the `pandas` object storing the dataset?

In [12]:
df=dframe
type(df)

pandas.core.frame.DataFrame

Answer: TODO

> ### Question 4.  How many samples (i.e., rows) are in this dataset?

In [16]:
# TODO
df.shape
#Shape returns the number of rows and columns in the dataframe


(244, 8)

Answer: TODO

> ### Question 5.  How many variables (i.e., columns) are in this dataset?

In [17]:
df.columns

Index([u'day', u'time', u'name', u'gender', u'is_smoker', u'party', u'check',
       u'tip'],
      dtype='object')

Answer: TODO

> ### Question 6.  Print the name of each column in the dataset, one name per line.

In [19]:
lst=df.columns
for i in lst:
    print i

day
time
name
gender
is_smoker
party
check
tip


> ### Question 7.  Print the first two rows of the dataset to the console.  What does the output look like?

In [24]:
print df.head(n=2)
#this returns the first x rows of the dataframe


      day    time      name  gender  is_smoker  party  check   tip
0  Sunday  Dinner  Kimberly  Female      False      2  16.99  1.01
1  Sunday  Dinner  Nicholas    Male      False      3  10.34  1.66


Answer: TODO

> ### Question 8.  Extract the last 2 rows of the data frame and print them to the console.  What does the output look like?

In [27]:
print df.tail(n=2)


          day    time    name  gender  is_smoker  party  check   tip
242  Saturday  Dinner     Jon    Male      False      2  17.82  1.75
243  Thursday  Dinner  Brandi  Female      False      2  18.78  3.00


Answer: TODO

> ### Question 9.  Does the dataset contain any missing values?

In [30]:
df.isnull().sum()
# this returns the number of rows that are null for a given column


day          0
time         0
name         0
gender       0
is_smoker    0
party        0
check        0
tip          0
dtype: int64

Answer: No - The dataset does not contain any null values

> ### Question 10.  What can you say about the `is_smoker` variable?  I.e., will it bring any insights when analyzing the dataset?  What do you want to do with it?  (and do it...)

In [43]:
df.describe()
#The data set does not have any nulls, not lets analyze what the columns is_smoker looks like 
df.is_smoker.unique()
#looking at the unique values of is_smoker we can tell that the only value in here is False
df.drop(['is_smoker'], axis = 1, inplace = True)

day    time      name  gender  party  check   tip
0      Sunday  Dinner  Kimberly  Female      2  16.99  1.01
1      Sunday  Dinner  Nicholas    Male      3  10.34  1.66
2      Sunday  Dinner     Larry    Male      3  21.01  3.50
3      Sunday  Dinner    Joseph    Male      2  23.68  3.31
4      Sunday  Dinner    Janice  Female      4  24.59  3.61
..        ...     ...       ...     ...    ...    ...   ...
239  Saturday  Dinner     Kevin    Male      3  29.03  5.92
240  Saturday  Dinner    Sandra  Female      2  27.18  2.00
241  Saturday  Dinner      Carl    Male      2  22.67  2.00
242  Saturday  Dinner       Jon    Male      2  17.82  1.75
243  Thursday  Dinner    Brandi  Female      2  18.78  3.00

[244 rows x 7 columns]

In [44]:
df

day    time      name  gender  party  check   tip
0      Sunday  Dinner  Kimberly  Female      2  16.99  1.01
1      Sunday  Dinner  Nicholas    Male      3  10.34  1.66
2      Sunday  Dinner     Larry    Male      3  21.01  3.50
3      Sunday  Dinner    Joseph    Male      2  23.68  3.31
4      Sunday  Dinner    Janice  Female      4  24.59  3.61
..        ...     ...       ...     ...    ...    ...   ...
239  Saturday  Dinner     Kevin    Male      3  29.03  5.92
240  Saturday  Dinner    Sandra  Female      2  27.18  2.00
241  Saturday  Dinner      Carl    Male      2  22.67  2.00
242  Saturday  Dinner       Jon    Male      2  17.82  1.75
243  Thursday  Dinner    Brandi  Female      2  18.78  3.00

[244 rows x 7 columns]

Answer: TODO

> ### Question 11.  For which week days does the dataset has data for?

In [48]:
df.columns
df.day.unique()

array(['Sunday', 'Saturday', 'Thursday', 'Friday'], dtype=object)

Answer: The data frame has data for Sunday/Saturday/Thursday/Friday

> ### Question 12.  How often was the bistro patronized for each week day?

(check `.value_counts()`; it could come in handy)

(http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html)

In [57]:
df.day.value_counts()
#This functions is supposed to returns the number of times a variable has been used in the dataframe for a given column


Saturday    87
Sunday      76
Thursday    62
Friday      19
Name: day, dtype: int64

Answer: TODO

> ### Question 13.  How much tip did waiters collect for each week day?

In [65]:
df.groupby('day')['tip'].sum()
#this returns the sum of all the tips collected in a day

day
Friday       51.96
Saturday    260.40
Sunday      247.39
Thursday    171.83
Name: tip, dtype: float64

In [60]:
df.columns



Index([u'day', u'time', u'name', u'gender', u'party', u'check', u'tip'], dtype='object')

Answer: TODO

> ### Question 14.  What is the average tip per check (in absolute \$) for each week day?

In [67]:
df.groupby('day')['tip'].mean()
#returns of the mean of the tip

day
Friday      2.734737
Saturday    2.993103
Sunday      3.255132
Thursday    2.771452
Name: tip, dtype: float64

In [69]:
#doing a rlough check to see if the calculation is correct
df.groupby('day')['tip'].count()

day
Friday      19
Saturday    87
Sunday      76
Thursday    62
Name: tip, dtype: int64

Answer: TODO

> ### Question 15.  What is the average tip per check (as a percentage of the check) for each week day?

In [96]:
# TODO
i = df.groupby('day')['tip'].sum()
type(i)
j = df.groupby('day')['check'].sum()
#taking the ratio of tip and the overall check
i/j*100


day
Friday      15.944519
Saturday    14.642375
Sunday      15.203791
Thursday    15.673201
dtype: float64

Answer: TODO

> ### Question 16.  Are there any name in common between male and female patrons?  (E.g., `Chris` can refer to either a man or a woman)

(check `numpy.intersect1d()`; it could come in handy)

(https://docs.scipy.org/doc/numpy/reference/generated/numpy.intersect1d.html)

In [130]:
df.columns
dfMale = df[df['gender'] == "Male"]
dfFemale = df[df['gender'] <> "Male"]
dfFemale.name
np.intersect1d(dfFemale.name,dfMale.name)
df.name.value_counts()
#np.interesectid(df.name[upper('gender')='MALE'], df.name[upper('gender')='FEMALE'])

Casey       8
David       8
James       5
Mary        4
Carl        4
           ..
Walter      1
Chandler    1
Reyes       1
Matthew     1
Sarah       1
Name: name, Length: 181, dtype: int64

Answer: TODO

> ### Question 17.  If no patrons share the same name, how many unique patrons are in the dataset?

In [135]:
len(df.name.unique())


181

Answer: 181

> ### Question 18.  How many times did `Kevin` patronized the bistro?  How about `Alice`?

In [144]:
len(df[df['name'] == "Kevin" ])+len(df[df['name'] == "Alice" ])

6

Answer: TODO

> ### Question 19.  Who are the top 3 female and male patrons?

In [163]:
k=dfFemale.name.value_counts().head(n=3)
k
j=dfMale.name.value_counts().head(n=3)
j

David    8
James    5
Casey    5
Name: name, dtype: int64

Answer: TODO

> ### Question 20.  Who's the best tipper (as a fraction of all tips over all check totals)?  Who's the worst?  How many times did they patronize the bistro?

- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)
- (http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.sort_values.html)

In [193]:
k =df.groupby('name')['tip'].sum() / df.groupby('name')['check'].sum() 
j= df.groupby('name')['check'].count()
k.sort_values(ascending = False)
j['name'== 'Maryann']

1

Answer: TODO

In [194]:
print bool(0)

False


In [23]:
#fetch the iris dataset from pandas.
import numpy as np
from sklearn import datasets
iris = datasets.load_iris()
iris_X = iris.data
iris_y = iris.target
np.unique(iris_y)
type(iris_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [12]:
type(dbc)
dbc.target_names


array(['malignant', 'benign'], 
      dtype='|S9')